core/payment/src/processor.rs

```diff
@@ -1,203 +1,288 @@
 use ya_net::RemoteEndpoint;
 use ya_persistence::executor::DbExecutor;
+use ya_service_bus::typed::Endpoint;
 use ya_service_bus::{typed as bus, RpcEndpoint};
 
+fn driver_endpoint(driver: &str) -> Endpoint {
+    bus::service(driver::BUS_ID_PREFIX.to_string() + driver)
+}
+
+async fn validate_orders(
+    orders: &Vec<DbOrder>,
+    platform: &str,
+    payer_addr: &str,
+    payee_addr: &str,
+    amount: &BigDecimal,
+) -> Result<(), OrderValidationError> {
+    let mut total_amount = BigDecimal::zero();
+    for order in orders.iter() {
+        if &order.payment_platform != platform {
+            return OrderValidationError::platform(order, platform);
+        }
+        if &order.payer_addr != &payer_addr {
+            return OrderValidationError::payer_addr(order, payer_addr);
+        }
+        if &order.payee_addr != &payee_addr {
+            return OrderValidationError::payee_addr(order, payee_addr);
+        }
+
+        total_amount += &order.amount.0;
+    }
+
+    if &total_amount != amount {
+        return OrderValidationError::amount(&total_amount, amount);
+    }
+
+    Ok(())
+}
+
 #[derive(Clone)]
 pub struct PaymentProcessor {
     db_executor: DbExecutor,
+    accounts: Arc<Mutex<HashMap<(String, String), (String, AccountMode)>>>, // (platform, address) -> (driver, mode)
+    drivers: Arc<Mutex<HashMap<String, String>>>,                           // driver -> platform
 }
 
 impl PaymentProcessor {
     pub fn new(db_executor: DbExecutor) -> Self {
-        Self { db_executor }
+        Self {
+            db_executor,
+            accounts: Default::default(),
+            drivers: Default::default(),
+        }
     }
 
-    async fn wait_for_payment(&self, invoice_id: &str) -> PaymentResult<PaymentConfirmation> {
-        loop {
-            let payment_status: PaymentStatus = bus::service(driver::BUS_ID)
-                .send(driver::GetPaymentStatus::from(invoice_id.to_string()))
-                .await
-                .unwrap()
-                .unwrap();
-            match payment_status {
-                PaymentStatus::Ok(confirmation) => return Ok(confirmation),
-                PaymentStatus::NotYet => tokio::time::delay_for(Duration::from_secs(5)).await,
-                _ => return Err(PaymentError::Driver(String::from("Insufficient funds"))),
+    pub async fn register_account(&self, msg: RegisterAccount) -> Result<(), RegisterAccountError> {
+        match self.drivers.lock().await.entry(msg.driver.clone()) {
+            Entry::Vacant(entry) => {
+                entry.insert(msg.platform.clone());
+            }
+            Entry::Occupied(entry) if entry.get() != &msg.platform => {
+                return Err(RegisterAccountError::Other(format!(
+                    "Driver {} is registered as handling platform {}",
+                    msg.driver,
+                    entry.get()
+                )))
             }
+            _ => {}
         }
+
+        match self
```

Comment:
It could be done with a single lock but that would require introducing some helper structure and thus more code. In what situation would it wait for a long time?

Reply:
I've extracted these two into a new struct `DriverRegistry` in https://github.com/golemfactory/yagna/pull/395/commits/027c2b771ea19249c56484de7e8bdd4118cab49e so that only one lock is used.